File: core/src/rustyice/editor/EditorPropertyPane.java
package rustyice.editor;

import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup;
import com.kotcrab.vis.ui.widget.*;
import rustyice.editor.annotations.ComponentAccess;
import rustyice.editor.annotations.ComponentProperty;
import rustyice.editor.widgetbuilders.BoolWidgetBuilder;
import rustyice.editor.widgetbuilders.ColorWidgetBuilder;
import rustyice.editor.widgetbuilders.FloatWidgetBuilder;
import rustyice.editor.widgetbuilders.PropertyWidgetBuilder;
import rustyice.game.GameObject;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;

/**
 * @author gabek
 */
public class EditorPropertyPane {
    private VisScrollPane propertyTreeScroll;
    private VisTree propertyTree;

    private GameObject selectedObject;

    public EditorPropertyPane(){
        propertyTree = new VisTree();
        propertyTreeScroll = new VisScrollPane(propertyTree);
    }

    public WidgetGroup getRoot(){
        return propertyTreeScroll;
    }

    public void setSelected(GameObject selectedObject){
        if(this.selectedObject != selectedObject){
            this.selectedObject = selectedObject;
            propertyTree.clearChildren();


            propertyTree.add(buildCompEditNode(selectedObject));
            HashSet<Object> compSet = new HashSet<>();

            for(Method method: selectedObject.getClass().getMethods()){
                if(getInheritedAnnotation(selectedObject.getClass(), method, ComponentAccess.class) != null){
                    try {
                        Object comp = method.invoke(selectedObject);
                        if(compSet.add(comp)){
                            propertyTree.add(buildCompEditNode(comp));
                        }
                    } catch (IllegalAccessException | InvocationTargetException e) {
                        e.printStackTrace();
                    }
                }
            }
        }
    }

    private <T extends Annotation> T getInheritedAnnotation(Class cls, Method mth, Class<T> anno){
        T out = mth.getAnnotation(anno);

        //check superclasses
        Class superclass = cls.getSuperclass();
        while (superclass != null && out == null){
            for(Method method: superclass.getDeclaredMethods()){
                if(method.getName().equals(mth.getName())
                        && Arrays.equals(method.getParameterTypes(), mth.getParameterTypes())
                        && method.getAnnotation(anno) != null){
                    out = method.getAnnotation(anno);
                    break;
                }
            }
            superclass = superclass.getSuperclass();
        }

        return out;
    }

    private VisTree.Node buildCompEditNode(Object comp) {

        VisTree.Node node = new VisTree.Node(new VisLabel(comp.getClass().getSimpleName()));
        node.setSelectable(false);

        HashMap<String, PropertyWidgetBuilder> propertyMap = new HashMap<>();

        for(Method method: comp.getClass().getMethods()){
            ComponentProperty propertyAnno = getInheritedAnnotation(comp.getClass(), method, ComponentProperty.class);

            if(propertyAnno != null){
                String title = propertyAnno.title();

                if(ColorWidgetBuilder.isSetter(method)) {
                    propertyMap.computeIfAbsent(title, s -> new ColorWidgetBuilder(comp, title)).setSetter(method);
                } else if(ColorWidgetBuilder.isGetter(method)){
                        propertyMap.computeIfAbsent(title, s -> new ColorWidgetBuilder(comp, title)).setGetter(method);
                }
                else if(FloatWidgetBuilder.isSetter(method)){
                    propertyMap.computeIfAbsent(title, s -> new FloatWidgetBuilder(comp, title)).setSetter(method);
                } else if(FloatWidgetBuilder.isGetter(method)) {
                    propertyMap.computeIfAbsent(title, s -> new FloatWidgetBuilder(comp, title)).setGetter(method);
                }
                else if(BoolWidgetBuilder.isSetter(method)){
                    propertyMap.computeIfAbsent(title, s -> new BoolWidgetBuilder(comp, title)).setSetter(method);
                } else if(BoolWidgetBuilder.isGetter(method)){
                    propertyMap.computeIfAbsent(title, s -> new BoolWidgetBuilder(comp, title)).setGetter(method);
                }
            }

        }

        for (PropertyWidgetBuilder propertyWidgetBuilder : propertyMap.values()) {
            VisTree.Node widgetNode = new VisTree.Node(propertyWidgetBuilder.buildWidgets());
            widgetNode.setSelectable(false);
            node.add(widgetNode);
        }

        return node;
    }

}

File: core/src/rustyice/editor/EditorPropertyPane.kt
package rustyice.editor

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.scenes.scene2d.ui.Tree
import com.badlogic.gdx.scenes.scene2d.ui.WidgetGroup
import com.kotcrab.vis.ui.widget.VisLabel
import com.kotcrab.vis.ui.widget.VisScrollPane
import com.kotcrab.vis.ui.widget.VisTree
import rustyice.editor.annotations.ComponentAccess
import rustyice.editor.annotations.ComponentProperty
import rustyice.editor.widgetbuilders.*
import rustyice.game.GameObject
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.KMutableProperty
import kotlin.reflect.KProperty
import kotlin.reflect.memberProperties

/**
 * @author gabek
 */
class EditorPropertyPane {
    private val propertyTreeScroll: VisScrollPane
    private val propertyTree: VisTree

    private var selectedObject: GameObject? = null
    private val validatorList: List<PropertyValidator>

    val root: WidgetGroup get() = propertyTreeScroll

    init {
        propertyTree = VisTree()
        propertyTreeScroll = VisScrollPane(propertyTree)

        val mValList = ArrayList<PropertyValidator>()
        mValList.add(TypePropertyValidator(Float::class, {FloatWidgetBuilder()}))
        mValList.add(TypePropertyValidator(Boolean::class, {BoolWidgetBuilder()}))
        mValList.add(TypePropertyValidator(Color::class, {ColorWidgetBuilder()}))

        validatorList = mValList
    }

    fun setSelected(selectedObject: GameObject){
        if(this.selectedObject != selectedObject){
            this.selectedObject = selectedObject
            propertyTree.clearChildren()

            propertyTree.add(buildCompEditNode(selectedObject))
            val compSet = HashSet<Any>()

            val c = selectedObject.javaClass.kotlin
            for(prop in c.members){
                if(prop is KProperty) {
                    if (getInheritedAnnotation(c, prop, ComponentAccess::class) != null) {
                        val comp = prop.call(selectedObject)
                        if (comp != null && compSet.add(comp)) {
                            propertyTree.add(buildCompEditNode(comp))
                        }
                    }
                }
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun <A: Annotation> getInheritedAnnotation(cls: KClass<*>, prop: KProperty<*>, anno: KClass<A>): A?{
        for(a in prop.annotations){
            if(a.annotationClass == anno){
                return a as A
            }
        }
        return null

        //check superclasses
        //var superclass = cls.
        //while (superclass != null && out == null){
        //    for(Method method: superclass.getDeclaredMethods()){
        //        if(method.getName().equals(mth.getName())
        //                && Arrays.equals(method.getParameterTypes(), mth.getParameterTypes())
        //                && method.getAnnotation(anno) != null){
        //            return true
        //        }
        //    }
        //    superclass = superclass.getSuperclass();
        //}

    }


    private fun buildCompEditNode(comp: Any): Tree.Node {
        val kClass = comp.javaClass.kotlin

        val node = Tree.Node(VisLabel(kClass.simpleName))
        node.isSelectable = false

        val propertyMap = ArrayList<PropertyWidgetBuilder>()

        for(prop in kClass.memberProperties){
            if(prop is KMutableProperty<*>) {
                val propertyAnno = getInheritedAnnotation(kClass, prop, ComponentProperty::class)

                if (propertyAnno != null) {
                    val title = prop.name
                    for(validator in validatorList){
                        if(validator.isInstance(prop)){
                            propertyMap.add(validator.widgetBuilder().init(title, comp, prop))
                        }
                    }
                }
            }
        }

        for (propertyWidgetBuilder in propertyMap) {
            val widgetNode = Tree.Node(propertyWidgetBuilder.buildWidgets())
            widgetNode.isSelectable = false
            node.add(widgetNode)
        }

        return node
    }

}
